File: 307-range-sum-query-mutable/307-range-sum-query-mutable.java
class NumArray {

    private final int[] nums;
    private int sum = 0;
    private int nums_sum_limit_length = 0;

    public NumArray(int[] nums) {
        this.nums = nums;
        for (int num : nums) {
            this.sum += num;
        }
        this.nums_sum_limit_length = nums.length / 2;
    }
    
    // index의 값을 val로
    public void update(int index, int val) {
        int change_ago_index = nums[index];
        this.sum -= change_ago_index;
        this.sum += val;
        this.nums[index] = val;
    }
    
    // left부터 right 인덱스 까지의 합을 리턴 
    public int sumRange(int left, int right) {
        int result = 0;

        if ((right - left) > this.nums_sum_limit_length) {
            result = sum;

            for (int i = 0; i < left; ++i) {
                result -= this.nums[i];
            }
            for (int i = right + 1; i < this.nums.length; ++i) {
                result -= this.nums[i];
            }

            return result;
        }

        for (int i = left; i <= right; ++i) {
            result += this.nums[i];
        }
        return result;
    }
}

File: 307-range-sum-query-mutable/307-range-sum-query-mutable.py
# Leet Code Solution
class NumArray:

    nums = []
    sum_value = 0
    nums_len  = 0
    
    def __init__(self, nums: List[int]):
        self.nums = nums
        self.sum_value = sum(nums)
        self.nums_len = len(nums)

    def update(self, index: int, val: int) -> None:
        self.sum_value -= self.nums[index]
        self.nums[index] = val
        self.sum_value += self.nums[index]

    def sumRange(self, left: int, right: int) -> int:
        if right - left > self.nums_len // 2:
            answer = sum(self.nums[:left]) + sum(self.nums[right + 1:])
            return self.sum_value - answer
        else:
            return sum(self.nums[left: right + 1])


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# obj.update(index,val)
# param_2 = obj.sumRange(left,right)


